save-agent/src/nativeMain/kotlin/org/cqfn/save/agent/SaveAgent.kt

```diff
@@ -0,0 +1,155 @@
+@file:Suppress("PACKAGE_NAME_INCORRECT_PATH")
+
+package org.cqfn.save.agent
+
+import com.benasher44.uuid.uuid4
+import io.ktor.client.HttpClient
+import io.ktor.client.features.HttpTimeout
+import io.ktor.client.features.json.JsonFeature
+import io.ktor.client.features.json.serializer.KotlinxSerializer
+import io.ktor.client.request.accept
+import io.ktor.client.request.post
+import io.ktor.http.ContentType
+import io.ktor.http.contentType
+import io.ktor.util.KtorExperimentalAPI
+import platform.posix.system
+
+import kotlin.native.concurrent.AtomicReference
+import kotlinx.atomicfu.atomic
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.modules.SerializersModule
+
+/**
+ * A main class for SAVE Agent
+ */
+@OptIn(KtorExperimentalAPI::class)
+class SaveAgent(private val config: AgentConfiguration,
+                private val httpClient: HttpClient = HttpClient {
+                    install(JsonFeature) {
+                        serializer = KotlinxSerializer(Json {
+                            serializersModule = SerializersModule {
+                                // for some reason for K/N it's needed explicitly, at least for ktor 1.5.1, kotlin 1.4.21
+                                contextual(HeartbeatResponse::class, HeartbeatResponse.serializer())
+                            }
+                        })
+                    }
+                    install(HttpTimeout) {
+                        requestTimeoutMillis = config.requestTimeoutMillis
+                    }
+                }
+) {
+    /**
+     * The current [AgentState] of this agent
+     */
+    val state = AtomicReference(AgentState.IDLE)
+    private val id = uuid4().toString()
+    private val isStopped = atomic(false)
+    private var saveProcessJob: Job? = null
+
+    /**
+     * @return Unit
+     */
+    suspend fun start() = coroutineScope {
+        println("Starting agent")
+        val heartbeatsJob = launch { startHeartbeats() }
+        heartbeatsJob.join()
+        println("Heartbeats job is done")
```

Comment:
logic for start is overloaded - this print should be done inside the corresponding function